### PR TITLE
CoAP: Random generate ID the first time coap_next_id() is called

### PR DESCRIPTION
**Pull request by tautologyclub**
This is more in accordance with CoAP recommendations (see
https://tools.ietf.org/html/draft-ietf-core-coap-18, section 4.4)

"It is strongly recommended that the initial value of the variable (e.g.,
on startup) be randomized, in order to make successful off-path attacks
on the protocol less likely."

Signed-off-by: Benjamin Lindqvist <benjamin.lindqvist@endian.se>

### CODE REVIEW

**Comment by tautologyclub** [comment]
According to codeowners, it looks like @rveerama1 might be interested?

**Review by rveerama1 — APPROVED** [review]
LGTM

File: include/net/coap.h
Comment (jukkar) [inline include/net/coap.h]: The atomic type should `atomic_t` instead of int, even thou in practice they are the same.

File: include/net/coap.h
Comment (jukkar) [inline include/net/coap.h]: The cast does no harm here so we could leave it, but is not really needed and could be omitted.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: include/net/coap.h
Comment (jukkar) [inline include/net/coap.h]: Please follow the commit message style for networking components. So the subject should read `net: coap: ....`

File: include/net/coap.h
Comment (tautologyclub) [reply include/net/coap.h]: I tried to be super explicit because I hear MISRA certs are in the pipeline.

File: include/net/coap.h
Comment (tautologyclub) [reply include/net/coap.h]: done

File: include/net/coap.h
Comment (tautologyclub) [reply include/net/coap.h]: done

**Review by jukkar — APPROVED** [review]

**Comment by tautologyclub** [comment]
@jukkar do you want me to take care of backporting? It's a one liner

**Comment by tautologyclub** [comment]
Not sure what the procedure is though

**Comment by jukkar** [comment]
> do you want me to take care of backporting?

There is a bot that will do the backporting automatically if the PR can be merged without conflicts. Lets see after this is merged whether the bot can do that, if it cannot, then manual backporting is needed.

**Comment by tautologyclub** [comment]
@jukkar It won't - sys/slist.h has changed to misc/slist.h

**Comment by jukkar** [comment]
> It won't - sys/slist.h has changed to misc/slist.h

Ok, then manual backporting is needed. But before we get that far, the sanity reported issues need to be solved, there seems to be timeout for coap test.

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/51029/4/tests

**Comment by tautologyclub** [comment]
I noticed an unexpected timeout now too where the entropy driver seems to get stuck. This is on nrf52840. Does not seem consistently reproducible. Will try to catch a backtrace.

**Comment by jukkar** [comment]
> Any idea how to reproduce the CI test locally?

I usually do this `sanitycheck -T x/y` if I want to check if some tests in a directory x/y fail.
Other options is for example `sanitycheck -t net`. One can also add `--all` option to sanitycheck if needed.
There is also `scripts/ci/run_ci.sh` that is run by CI but I have not used it myself.

**Comment by tautologyclub** [comment]
This particular suite seems to have some dependencies. On net ifs of some particular fashion.

**Comment by tautologyclub** [comment]
Ehm.

I can reproduce this now. And I'm not sure this code doesn't produce undefined behavior. Same for the original function. Maybe someone can help me decipher what the C standard says about modifiable static variables inside static inline functions?

`coap_next_id()` becomes completely erratic when you add the `seeded` check and my only guess right now is due to undefined behaviour caused by static variables in a static inline function. Input very welcome however. @jukkar  @rveerama1

**Comment by tautologyclub** [comment]
Updates: I made the function a "normal" non-inlined function in `coap.c`, just included the "normal" prototype in `coap.h` and suddenly everything seems to work without a hitch. Will push the code soon

**Comment by jukkar** [comment]
> I made the function a "normal" non-inlined function in `coap.c`, just included the "normal" prototype in `coap.h` and suddenly everything seems to work without a hitch. Will push the code soon

Good catch! Indeed the static variable and inline function are bad combination.

**Comment by marc-hb** [comment]
> Good catch! Indeed the static variable and inline function are bad combination.

I could think of some corner case(s) where you want to avoid the stack growing too much and you know for sure there won't be any concurrent use but... yeah, that would deserve a big fat comment at the very least. 

#19380

**Comment by tautologyclub** [comment]
I forgot to include the atomic_cas in the new commit for some reason. I don't think it does much but might as well use it, right?

File: subsys/net/lib/coap/coap.c
Comment (jukkar) [inline subsys/net/lib/coap/coap.c]: As you mentioned, it would be better if this is atomic_t

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/lib/coap/coap.c
Comment (tautologyclub) [reply subsys/net/lib/coap/coap.c]: @jukkar Had a busy last week so apologies for the latency. It's pushed now.

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
This definitely needs to be squashed. And why @mike-scott isn't in auto-reviewers for this?

**Review by pfalcon — COMMENTED** [review]
(now again as a review comment)

This definitely needs to be squashed. And why @mike-scott isn't in auto-reviewers for this?

**Comment by jukkar** [comment]
> This definitely needs to be squashed.

Not really, there are two different changes so those can be kept in separate commits.

**Review by mike-scott — APPROVED** [review]
LGTM.  Thanks for this change.

**Comment by mike-scott** [comment]
Agree with @jukkar  -- no need to squash these 2  commits.  They deal with separate changes.

**Comment by pfalcon** [comment]
> Not really, there are two different changes so those can be kept in separate commits.

That's not what I see. I see that first commit mistakenly patches static inline function with a lot of new functionality, which generally doesn't belong to a static inline function. Then second commit fixes that mistake by converting the static inline to normal function. Normally, this should be done as one step. And as a general rule of thumb, if some code in patch series patches the same code twice, it's good indication that it those commites should be squashed together, as intermediate changes are usually just noise.

**Comment by mike-scott** [comment]
> > Not really, there are two different changes so those can be kept in separate commits.
> 
> That's not what I see. I see that first commit mistakenly patches static inline function with a lot of new functionality, which generally doesn't belong to a static inline function. Then second commit fixes that mistake by converting the static inline to normal function. Normally, this should be done as one step. And as a general rule of thumb, if some code in patch series patches the same code twice, it's good indication that it those commites should be squashed together, as intermediate changes are usually just noise.

What I see more often is the 2nd commit in this series landing first as a "prep" patch. The author knows that the inline static changes are needed for the next change, so they introduce it first.   That's more of a personal style issue, however, and I try to be lenient in that area.

**Comment by tautologyclub** [comment]
If the first commit INTRODUCED the bug then squashing would be obviously right. It didn't though, it just made a preexisting bug more noticeable. There were two problems with the coap code and now there are two fix commits. Can't see the big deal.

**Comment by jukkar** [comment]
It is not very obvious why the sanity checks fail as the changes are quite simple. @tautologyclub, any idea why we get timeouts?

**Comment by tautologyclub** [comment]
@jukkar Is it the same behavior as you noticed before the second commit? Have a hard time parsing the shippable report - it doesn't seem to start running any tests? The only thing I can think of is some problem with the qemu/x86 entropy driver?

**Comment by jukkar** [comment]
Investigated this a bit. The tests pass if I make the following three test functions return TC_PASS in `tests/net/lib/coap/src/main.c`
```
	{ "Test block sized 1 transfer", test_block1_size, },
	{ "Test block sized 2 transfer", test_block2_size, },
	{ "Test retransmission", test_retransmit_second_round, },
```

Tested with
`sanitycheck -p qemu_x86 -s tests/net/lib/coap/net.coap`
`sanitycheck -p mps2_an385 -s tests/net/lib/coap/net.coap`

@tautologyclub Note that you should rebase your tree against latest master if you want to try the above sanitycheck scripts, the platform whitelisting was removed from `tests/net/lib/coap/testcase.yaml` recently but your tree seems to still have it.

**Comment by tautologyclub** [comment]
@jukkar This is the first time I try the sanitycheck script. This is what happens (after rebase onto master):

```
~/zephyrproject/zephyr:(HEAD detached from endian_github/initialize-coap-message-id)
$ scripts/./sanitycheck -p qemu_x86 -s tests/net/lib/coap/net.coap                           
JOBS: 8
Building testcase defconfigs...
Filtering test cases...
0 tests selected, 137034 tests discarded due to filters

0 of 0 tests passed with 0 warnings in 2 seconds
```

**Comment by tautologyclub** [comment]
@jukkar I've ran the tests using qemu_cortex_m3 locally now and I think I see the problem. Will validate and patch if I can, asap

**Comment by tautologyclub** [comment]
So the thing is that `atomic_cas` only returns `true` if the value CHANGES according to the comments in `sys/atomic.h`, resulting in an infinite while loop. I actually thought it returned `true` if `target_value == new_value` and `false` only if the atomicity requirement was broken (e.g. it was interrupted or atomic operations were temporarily unavailable). So the correct version would really be

	if (atomic_cas(&seeded, 0, 1)) {
		message_id = (u16_t)sys_rand32_get();
	}

right?

However, isn't it pretty weird that this does not happen when running on target? (nrf52840) Both target and qemu_m3 has `CONFIG_ATOMIC_OPERATIONS_BUILTIN=y`. Do the builtins behave differently...?

File: subsys/net/lib/coap/coap.c
Comment (jukkar) [inline subsys/net/lib/coap/coap.c]: This certainly works but there is still possibility that one CoAP message would have message id 1. For example if the `sys_rand32_get()` would block then it could cause some other thread to run, then that other thread could possible get message_id = 1 as a result. Chance for this is very minimal.

One solution for that could be that we introduce `coap_init()` function that could be called from `services_init()` which is in `subsys/net/ip/net_core.c`, and which could do the message_id init.

Comment (tautologyclub) [reply subsys/net/lib/coap/coap.c]: Agree that's way more idiomatic

Comment (tautologyclub) [reply subsys/net/lib/coap/coap.c]: I interpreted the "standard" approach to the best of my ability, but feel free to place some extra attention on `net_core.c`. There doesn't seem to be a consensus best practice for conditionally NOPing functions so I just sort of picked one.

File: subsys/net/ip/net_core.c
Comment (jukkar) [inline subsys/net/ip/net_core.c]: Please add this to coap.h, and include coap.h in this file. We can also call it coap_init(), the net_ prefix is not really needed in this case (as the coap API does not have it).

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks proper, one nit pick for the func prototype and naming.

File: subsys/net/ip/net_core.c
Comment (jukkar) [reply subsys/net/ip/net_core.c]: Also, should we add a check to coap_init() that it cannot be initialized multiple times?

Comment (tautologyclub) [reply subsys/net/ip/net_core.c]: That's the reason I didn't include it in the header -- because it's not supposed to be publicly exposed - that would probably mean people peruse `coap.h` and infer that THE APP should be calling `coap_init()`. I felt declaring it `extern` was preferable to exposing internal stuff in public headers. This way, only `net_core` even knows it exists.

Comment (tautologyclub) [reply subsys/net/ip/net_core.c]: (this would also mean the multiple calls check would be superfluous)

Comment (jukkar) [reply subsys/net/ip/net_core.c]: > hat would probably mean people peruse `coap.h` and infer

Sure, that is certainly a valid concern. We can also place the init declaration to `subsys/net/ip/net_private.h` as there already other init function declarations.

Comment (tautologyclub) [reply subsys/net/ip/net_core.c]: Yeah that sounds better. How's this?

**Review by jukkar — APPROVED** [review]
LGTM

**Pull request merged**
